File: 0006-zigzag-conversion/0006-zigzag-conversion.go
func convert(s string, numRows int) string {
    rows := make([]strings.Builder, numRows+1)
    
    idx := 0
    for idx < len(s) {
        for i := 0; i < numRows && idx < len(s); i++ {
            rows[i].WriteByte(s[idx])
            idx++
        }
        for j := numRows-2; 0 < j && idx < len(s); j-- {
            rows[j].WriteByte(s[idx])
            idx++
        }
    }
    
    var zigzag strings.Builder
    for _, r := range rows {
        zigzag.WriteString(r.String())
    }
    return zigzag.String()
}

File: 0006-zigzag-conversion/0006-zigzag-conversion.py
class Solution:
    def convert(self, s: str, numRows: int) -> str:
        rows = {}
        chars = list(s)
        
        while chars:
            for i in range(numRows):
                if chars:
                    rows[i] = rows.get(i, "") + chars.pop(0)
            for i in range(numRows-2, 0, -1):
                if chars:
                    rows[i] = rows.get(i, "") + chars.pop(0)
            
        zigzag = ""
        for line in rows.values():
            zigzag += line
            
        return zigzag    